Add separate `minimize_disabled` and `maximize_disabled` window flags.

bruvzg opened the pull request:
Adds dedicated `minimize_disabled` and `maximize_disabled` window flags to disable minimize and maximize buttons. Previously maximized button was automatically disabled when resizable set to false, minimize was never disabled.

Implemented on macOS and Windows.

Note: flags are set on X11 as well, but it seems like X11 ignore it (might be window manager spceific). Not sure if it's possible to set it on Wayland.

Alternative to https://github.com/godotengine/godot/pull/105066 (disabled minimize, and only on Windows).

bruvzg committed:
Add separate `minimize_disabled` and `maximize_disabled` window flags.

KoBeWi commented on doc/classes/Window.xml:
```diff
@@ -624,10 +624,18 @@
 		</member>
+		<member name="maximize_disabled" type="bool" setter="set_flag" getter="get_flag" default="false">
+			If [code]true[/code], the [Window] maximize button is disabled.
+			[b]Note:[/b] This property is implemented only on macOS.
```
```suggestion
			[b]Note:[/b] This property is implemented only on macOS and Windows.
```
Same below.

KoBeWi commented on scene/gui/popup.cpp:
```diff
@@ -223,6 +223,8 @@ Popup::Popup() {
 	set_transient(true);
 	set_flag(FLAG_BORDERLESS, true);
 	set_flag(FLAG_RESIZE_DISABLED, true);
+	set_flag(FLAG_MINIMIZE_DISABLED, true);
+	set_flag(FLAG_MAXIMIZE_DISABLED, true);
```
Popups are borderless.
Did you mean it for dialogs?

bruvzg replied:
Dialogs are resizable, but maximizing and minimizing them is not useful, so it's probably ok to set it as default for `AcceptDialog`. It can be overridden for specific dialogs if it is actually desired.

KoBeWi commented:
Windows has this unique behavior where disabling either minimize or maximize will make them look disabled, but if you disable both, they will disappear completely. Might be worth documenting somewhere.

bruvzg commented:
> idk if it's the same for mac.

No, it's just grayed out, but we probably can hide them entirely to make it more consistent.

bruvzg commented:
> disabling either minimize or maximize will make them look disabled, but if you disable both, they will disappear completely

Modified macOS code to do the same, and added a note to the docs.

Also added these flags as default to `AcceptDialog`.

kitbdev commented:
Tested and looks good.

- supersedes #93861
- supersedes #85911
- closes https://github.com/godotengine/godot-proposals/issues/8468
- closes https://github.com/godotengine/godot-proposals/issues/10092
- fixes https://github.com/godotengine/godot/issues/92047
- fixes https://github.com/godotengine/godot/issues/80309
- fixes https://github.com/godotengine/godot/issues/62668
- related https://github.com/godotengine/godot/issues/37377

kitbdev commented on editor/plugins/animation_state_machine_editor.cpp:
```diff
@@ -1946,6 +1946,8 @@ AnimationNodeStateMachineEditor::AnimationNodeStateMachineEditor() {
 
 	delete_window = memnew(ConfirmationDialog);
 	delete_window->set_flag(Window::FLAG_RESIZE_DISABLED, true);
+	delete_window->set_flag(Window::FLAG_MINIMIZE_DISABLED, true);
+	delete_window->set_flag(Window::FLAG_MAXIMIZE_DISABLED, true);
```
ConfirmationDialog is an AcceptDialog, so these are already true by default.
Also I think the delete_window is never used. #75759 removed the only time it is shown.

Mickeon commented on doc/classes/Window.xml:
```diff
@@ -712,7 +722,7 @@
 			If [code]true[/code], the [Window] can't be focused nor interacted with. It can still be visible.
 		</member>
 		<member name="unresizable" type="bool" setter="set_flag" getter="get_flag" default="false">
-			If [code]true[/code], the window can't be resized. Minimize and maximize buttons are disabled.
+			If [code]true[/code], the window can't be resized.
```
I don't believe many will be bothered by this change. But, because this is a change in behaviour nonetheless, it does break user's expectations. So... it breaks compatibility?

bruvzg replied:
Yes. But it was not working as documented, only maximize button was disabled.

Mickeon commented on doc/classes/Window.xml:
```diff
@@ -624,10 +624,20 @@
 			[b]Note:[/b] This property will be ignored if the value is lower than [member min_size].
 		</member>
+		<member name="maximize_disabled" type="bool" setter="set_flag" getter="get_flag" default="false">
+			If [code]true[/code], the [Window] maximize button is disabled.
```
```suggestion
			If [code]true[/code], the [Window]'s maximize button is disabled.
```

Mickeon commented on doc/classes/Window.xml:
```diff
@@ -624,10 +624,20 @@
 			[b]Note:[/b] This property will be ignored in favor of [method get_contents_minimum_size] if [member wrap_controls] is enabled and if its size is bigger.
 		</member>
+		<member name="minimize_disabled" type="bool" setter="set_flag" getter="get_flag" default="false">
+			If [code]true[/code], the [Window] minimize button is disabled.
```
```suggestion
			If [code]true[/code], the [Window]'s minimize button is disabled.
```

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -948,6 +948,12 @@
 			[b]Note:[/b] This setting only affects the exported project, or when the project is run from the command line. In the editor, the value of [member EditorSettings.run/window_placement/screen] is used instead.
 		</member>
+		<member name="display/window/size/maximize_disabled" type="bool" setter="" getter="" default="false">
+			Main window maximize button is disabled.
```
```suggestion
			If [code]true[/code], the main window's maximize button is disabled.
```

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -948,6 +948,12 @@
+			Main window maximize button is disabled.
+		</member>
+		<member name="display/window/size/minimize_disabled" type="bool" setter="" getter="" default="false">
+			Main window minimize button is disabled.
```
```suggestion
			If [code]true[/code], the main window's minimize button is disabled.
```

Mickeon reviewed (commented):
I don't know how to feel about the negative here. I gather that's how it's implemented at the OS level, but in Godot "`false` means enabled" tends to be avoided. At the same time, though, the API is no stranger to this (`no_focus`).

Repiteo commented:
Thanks!
